compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -287,6 +287,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// When encountering a closure that captures variables, where a FnPtr is expected,
+    /// suggest a non-capturing closure
+    pub(in super::super) fn suggest_no_capture_closure(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+    ) {
+        if matches!(expected.kind(), ty::FnPtr(_)) {
+            if let ty::Closure(def_id, _) = found.kind() {
+                if let Some(upvars) = self.tcx.upvars_mentioned(def_id.clone()) {
+                    err.note("closures can only be coerced to `fn` types if they do not capture any variables");
+
+                    // Report only one upvar being captured to reduce the amount error messages
+                    // reported back to the user.
+                    let first_upvar = upvars.get_index(0);
+                    if let Some((var_hir_id, upvar)) = first_upvar {
+                        let var_name = self.tcx.hir().name(*var_hir_id).to_string();
+                        let msg = format!("{} captured here", var_name);
+                        err.span_note(upvar.span, &msg);
+                    }
```

Comment:
Doing the following (note, needs formatting):

```suggestion
                    let spans_and_labels: Vec<(Span, String)> = upvars.iter().map(|(var_hir_id, upvar)| {
                        let var_name = self.tcx.hir().name(*var_hir_id).to_string();
                        let msg = format!("{} captured here", var_name);
                        (upvar.span, msg)
                    }).collect();
                    let multi_span: MultiSpan = spans_and_labels.iter().map(|(sp, _)| *sp).collect::<Vec<_>>().into();
                    for (sp, label) in spans_and_labels {
                        multi_span.push_span_label(sp, label);
                    }
                    err.span_note(multi_span, "closures can only be coerced to `fn` types if they do not capture any variables");
```

would produce the following output:

```
note: closures can only be coerced to `fn` types if they do not capture any variables
  --> $DIR/closure-no-fn-1.rs:6:39
   |
LL |     let foo: fn(u8) -> u8 = |v: u8| { a += v; a };
   |                                       ^ `a` captured here
```

That code doesn't restrict itself to only one var (it would make sense to do maybe up to 4 or so), but in practice might not be terrible.